### PR TITLE
kubeadm - graduate kubelet-start phase

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR graduates phase kubelet-start in the kubeadm init workflow.

Following subcommands are automatically added to kubeadm by the PhaseRunner:

```
kubeadm init phase kubelet-start [flags]
```

Following alpha phase sub command are removed

```
kubeadm alpha phase kubelet config write-to-disk
kubeadm alpha phase kubelet write-env-file
```

**Which issue(s) this PR fixes** : 
Rif https://github.com/kubernetes/kubeadm/issues/1163

/sig cluster-lifecycle
/kind feature
/cc @neolit123 
/assign @timothysc 
@kubernetes/sig-cluster-lifecycle-pr-reviews 
@chuckha 
@rosti 
@ereslibre 

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: cmd/kubeadm/app/cmd/phases/kubelet.go [hunk 1]
```diff
@@ -19,12 +19,13 @@ package phases
 import (
 	"fmt"
 
+	"github.com/golang/glog"
 	"github.com/spf13/cobra"
-
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/kubelet.go]: leave this empty line here?

File: cmd/kubeadm/app/cmd/phases/kubelet.go [hunk 2]
```diff
@@ -111,72 +94,73 @@ var (
...
-			err := RunKubeletWriteEnvFile(cfgPath)
-			kubeadmutil.CheckErr(err)
-		},
+// NewKubeletStartPhase creates a kubeadm workflow phase that start kubelet on a node.
+func NewKubeletStartPhase() workflow.Phase {
+	return workflow.Phase{
+		Name:    "kubelet-start",
+		Short:   "Writes kubelet settings and (re)starts kubelet",
+		Long:    "Writes a file with KubeletConfiguration and an environment file with node specific kubelet settings, and then (re)starts kubelet.",
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/kubelet.go]: `(re)starts the kubelet.`

File: cmd/kubeadm/app/cmd/phases/kubelet.go [hunk 3]
```diff
@@ -111,72 +94,73 @@ var (
...
-	options.AddConfigFlag(cmd.Flags(), &cfgPath)
-	return cmd
 }
 
-// RunKubeletWriteEnvFile is the function that is run when "kubeadm phase kubelet write-env-file" is executed
-func RunKubeletWriteEnvFile(cfgPath string) error {
-	internalcfg, err := configutil.AnyConfigFileAndDefaultsToInternal(cfgPath)
-	if err != nil {
-		return err
+// runKubeletStart executes start kubelet logic.
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/kubelet.go]: `start-kubelet`?

File: cmd/kubeadm/app/cmd/phases/kubelet_test.go [hunk 4]
```diff
@@ -20,16 +20,13 @@ import (
 	"testing"
 
 	"github.com/spf13/cobra"
-
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/phases/kubelet_test.go]: hm, is this something the new gofmt does - i.e. removing these lines?

**Review by neolit123 — COMMENTED** [review]
LGTM, thanks @fabriziopandini 
added some minor comments.

this conflicts with @chuckha 's PR here:
https://github.com/kubernetes/kubernetes/pull/69985

in terms of when we want stop the kubelet.
i'm holding both this an his PR until we have a discussion about it. :[
/hold

File: cmd/kubeadm/app/cmd/phases/kubelet.go [hunk 5]
```diff
@@ -111,72 +94,73 @@ var (
...
+
+	// Write the kubelet configuration file to disk.
+	if err := kubeletphase.WriteConfigToDisk(data.Cfg().ComponentConfigs.Kubelet, data.KubeletDir()); err != nil {
+		return fmt.Errorf("error writing kubelet configuration to disk: %v", err)
+	}
+
+	// Try to start the kubelet service in case it's inactive
+	if !data.DryRun() {
+		glog.V(1).Infof("Starting the kubelet")
+		kubeletphase.TryStartKubelet()
```
Comment (chuckha) [inline cmd/kubeadm/app/cmd/phases/kubelet.go]: it's a little weird that this function only prints errors and doesn't return any errors.

kubeadm doesn't really have a way to fail out of this phase. If `TryStartKubelet` fails to start the Kubelet, a message is printed but nothing is returned.

This is probably ok and an existing limitation, but we could bail out of executing the phases earlier if this function returned an error.

File: cmd/kubeadm/app/cmd/phases/kubelet.go [hunk 6]
```diff
@@ -111,72 +94,73 @@ var (
...
 }
 
-// RunKubeletWriteEnvFile is the function that is run when "kubeadm phase kubelet write-env-file" is executed
-func RunKubeletWriteEnvFile(cfgPath string) error {
-	internalcfg, err := configutil.AnyConfigFileAndDefaultsToInternal(cfgPath)
-	if err != nil {
-		return err
+// runKubeletStart executes start kubelet logic.
+func runKubeletStart(c workflow.RunData) error {
+	data, ok := c.(kubeletStartData)
```
Comment (chuckha) [inline cmd/kubeadm/app/cmd/phases/kubelet.go]: what's the reason behind using the `kubeletStartData` interface instead of casting it to a `*initData`?

**Review by chuckha — COMMENTED** [review]
this is cool, i'm liking how the workflow/phases runner is coming together.

File: cmd/kubeadm/app/cmd/phases/kubelet.go [hunk 7]
```diff
@@ -111,72 +94,73 @@ var (
...
+	// Write env file with flags for the kubelet to use. We do not need to write the --register-with-taints for the master,
+	// as we handle that ourselves in the markmaster phase
+	// TODO: Maybe we want to do that some time in the future, in order to remove some logic from the markmaster phase?
+	if err := kubeletphase.WriteKubeletDynamicEnvFile(&data.Cfg().NodeRegistration, data.Cfg().FeatureGates, false, data.KubeletDir()); err != nil {
 		return fmt.Errorf("error writing a dynamic environment file for the kubelet: %v", err)
 	}
+
+	// Write the kubelet configuration file to disk.
+	if err := kubeletphase.WriteConfigToDisk(data.Cfg().ComponentConfigs.Kubelet, data.KubeletDir()); err != nil {
+		return fmt.Errorf("error writing kubelet configuration to disk: %v", err)
```
Comment (timothysc) [inline cmd/kubeadm/app/cmd/phases/kubelet.go]: We should really be using errors.Wrap when we refactor.  Non-blocking comment.

**Review by timothysc — APPROVED** [review]
/approve 

generally LGTM.

File: cmd/kubeadm/app/cmd/phases/kubelet.go (hunk 6 above)
Comment (timothysc) [reply cmd/kubeadm/app/cmd/phases/kubelet.go]: It's local to this phase.

File: cmd/kubeadm/app/cmd/phases/kubelet.go (hunk 1 above)
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/kubelet.go]: gofmt automatically removes those lines...

File: cmd/kubeadm/app/cmd/phases/kubelet.go (hunk 2 above)
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/kubelet.go]: done

File: cmd/kubeadm/app/cmd/phases/kubelet.go (hunk 3 above)
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/kubelet.go]: done

File: cmd/kubeadm/app/cmd/phases/kubelet_test.go (hunk 4 above)
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/kubelet_test.go]: yes...

File: cmd/kubeadm/app/cmd/phases/kubelet.go (hunk 5 above)
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/kubelet.go]: I agree, I'm opening an issue for changing this here and in the join workflow

File: cmd/kubeadm/app/cmd/phases/kubelet.go (hunk 7 above)
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/kubelet.go]: done

**Comment by fabriziopandini** [comment]
@neolit123  
@chuckha  reviewed this PR without showing conflicts with his PR (the same is valid for me the other way around). Could you remove /hold now?

**Comment by chuckha** [comment]
I need to rethink my approach so I will unhold this and I can work around it later

/hold cancel

**Comment by timothysc** [comment]
@fabriziopandini - integ tests are repeatedly failing.

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve

**Pull request merged**